Sleep rework, RTOS API for bare metal, wait deprecations

kjbracey opened the pull request:
### Description

A chunk of work bringing together connected themes.

* Decouple timed sleep code from RTOS idle
* Some refinements to that sleep code
* Separate C++ RTOS API from actual RTX implementation
* Implement bare metal subset of RTOS API using timed sleep code
* Deprecate `wait` APIs that actually sleep - only remaining ones spin and are safe from weird contexts.
* Small number of C RTOS APIs to allow basic sleep from C.

Details:
* `rtos/mbed_lib.json` moved down to `rtos/TARGET_CORTEX`, still with `name: "rtos"` so it's still checked by `MBED_CONF_RTOS_PRESENT`.
* new `rtos/mbed_lib.json` has `name: "rtos-api"`, so checked by `MBED_CONF_RTOS_API_PRESENT`, and accessible as `rtos/Semaphore.h` still.
* Basic C sleep and get_ms_count APIs in platform/mbed_thread.h, in case C++ RTOS API isn't present, or we need to sleep from C.
* Couple of new compatibility headers to smooth over RTOS/non-RTOS difference, eg include `mbed_rtos_types.h` instead of `cmsis_os2.h`  (cmsis_os2.h is not there in bare-metal, just the C++ RTOS APIs).

Points for discussion: 
* Directory structure, naming
* Final detail of naming/deprecations for `wait` APIs - deprecate all and split into `sleep` and `delay`?

<strike>TODO: Make event loop use the RTOS API for its sleep</strike>  done
TODO: Recheck exact changes made to sleep code, for discussion with Russ (GitHub diff isn't useful)

With this PR, you can take blinky, add `"requires" : [ "bare-metal" ]` to its mbed_app.json change its `wait_ms` to `ThisThread::sleep_for`, and you've now got a bare metal blinky that properly deep sleeps, with all the latency compensation stuff you get in the RTOS.  (For backwards compatibility, `wait_ms` remains non-sleeping in non-RTOS builds).

### Pull request type

    [ ] Fix
    [ ] Refactor
    [ ] Target update
    [X] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### Reviewers

@c1728p9, @sg-, @geky, @theotherjimmy, @bridadan, @bulislaw 

### Release Notes

* Subset of RTOS API now available in bare metal builds
* Full power save functionality now available in bare metal builds via RTOS API
* Consistent sleep vs wait API naming - wait() and wait_ms() deprecated

### Migration notes

* Use of `wait` and `wait_ms` should be reviewed to check whether sleeping is appropriate or not in the context the call occurs - if so, change to `ThisThread::sleep_for`/`thread_sleep_for`, else change to `wait_us`.

kjbracey committed:
Timed sleep rework

kjbracey committed:
Deprecate wait/wait_ms APIs

kjbracey committed:
CPU stats: strip more when disabled

kjbracey committed:
RTOS API for bare metal

Provide partial RTOS API for bare metal builds - things that
can be done in a single threaded environment.

Allows more code to work in both RTOS and bare metal builds without
change, and in particular gives easy access to the ability to
efficiently wait for something occurring in interrupt.

Available in bare-metal:
* ThisThread
* osThreadFlagsSet to set flags on main thread (can be set from IRQ)
* EventFlags (can be set from IRQ)
* Semaphores (can be released from IRQ)
* Mutex (dummy implementation)

Not useful:
* ConditionVariable (could only be signalled from 2nd thread)
* RtosTimer (calls in a second thread context)
* Thread

Unimplemented:
* Mail, Queue, MemoryPool

Possible future work:
* ConditionVariableCS to act as IRQ signalled ConditionVariable

kjbracey committed:
Minimal C thread API

kjbracey committed:
Make events use RTOS API

Switch from CMSIS-RTOS to mbed C++ API, which are available in bare
metal build.

Other minor tidies, like removing unnecessary volatile.

kjbracey committed:
CPU stats don't require low power ticker

kjbracey committed:
Replace deprecated wait calls

0xc0170 commented on rtos/rtos_c.h:
```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright (c) 2019 ARM Limited. All rights reserved.
```
as in the license guide, the file should contain the copyright, SPDX identifier and license text

0xc0170 commented on rtos/mbed_rtos_storage.h:
```diff
@@ -0,0 +1,33 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2006-2017 ARM Limited
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
```
Our files - we should use apache 2.0 (some files are MIT in this PR)

kjbracey replied:
I've updated the newly-created and moved files, but the bulk of the RTOS files remain MIT. Any reason they've not been changed? If there is, maybe the moved files shouldn't be changed.

0xc0170 commented on rtos/Mutex.cpp:
```diff
@@ -51,9 +53,9 @@ void Mutex::constructor(const char *name)
 }
 
-osStatus Mutex::lock(void)
+osStatus_t Mutex::lock(void)
```
@bulislaw was correct, this was changed in version 5 (`osStatus codes simplified and renamed to osStatus_t` 😞 ) but we kept the old one for compability in our RTOS API.

Do we have justification for this breakage?

kjbracey replied:
I think it did change anyway in RTX5 - as far as I can see in CMSIS 1/RTX4 `osStatus` used to be an enum, but now running RTX5 in this "CMSIS 1 compatibility mode", it's an `int32_t`. Whereas `osStatus_t` is a new enum. I think.

It wasn't an intentional change - I thought they were typedeffed to each other, so it was just making the code more modern-looking. I'll put it back.

0xc0170 commented on rtos/Semaphore.h:
```diff
@@ -98,8 +98,15 @@ class Semaphore : private mbed::NonCopyable<Semaphore> {
     mbed_rtos_storage_semaphore_t _obj_mem;
+#else
+    static bool semaphore_available(void *);
+    bool trywait();
```
is this new function - shouldn't it be `try_wait` ?

kjbracey replied:
Agree - it was being modelled after the badly-named `Mutex::trylock`.

I actually have another pending commit which didn't make it into this PR which deprecates `Semaphore::wait`.

> Introduce Semaphore::acquire methods
>     
> Deprecate wait() in favour of acquire(), try_acquire(), try_acquire_for() and try_acquire_until().
>     
> Brings Semaphore more into line with CMSIS-RTOS 2 (which uses "acquire"), itself (as it has "release"), and other classes having "try", "try for" and "try until".
>     
> Also steps away from vague "wait" term - the primary operation here is to acquire the semaphore, and this will of course sleep.

I'll fold that into the next version - it's another piece of removing "wait" ambiguity.

kjbracey replied:
Actually, will be a follow-up PR, as it has to change dozens of files to remove the deprecated `Semaphore::wait`s. But changed this function to be the same `try_acquire` it becomes later. (It's currently private in this PR, but becomes public).

0xc0170 commented on features/netsocket/NetworkInterfaceDefaults.cpp:
```diff
@@ -18,7 +18,7 @@
 #include "EthInterface.h"
 #include "WiFiInterface.h"
-#include "CellularBase.h"
+#include "CellularInterface.h"
```
Can this change be separated from this PR?

kjbracey replied:
Sure. No idea why it's even in here.

kjbracey replied:
Actually, it already went through as #10146.

0xc0170 reviewed (changes requested):
What is the use case for "Minimal C RTOS API" ?  There are only 3 functions defined currently.

rtos-api vs rtos?

bridadan commented on rtos/TARGET_CORTEX/mbed_lib.json:
```diff
@@ -0,0 +1,44 @@
+        "*": {
+            "target.boot-stack-size": "0x400"
+        },
+	    "MCU_NRF51": {
```
While this is being moved, perhaps you could correct this indent?

bridadan commented on platform/mbed_wait_api.h:
```diff
@@ -60,6 +62,10 @@ extern "C" {
+MBED_DEPRECATED_SINCE("mbed-os-5.13",
+                      "'wait'-named functions are now reserved for CPU-using delays that are safe to call from IRQ context. "
+                      "This call should be replaced by either a sleep call such as ThisThread::sleep_for (C++) or "
+                      "thread_sleep_for (C), or by a wait call such as wait_us(), depending on required behaviour.")
```
This message is a bit confusing to me. Is the goal to eventually remove `wait()` and `wait_ms()` with `wait_us()` only remaining as busy-wait implementation? Or will `wait()` and `wait_ms()` eventually transition to being a busy-wait implementation as well?

kjbracey replied:
Goal is to remove `wait` and `wait_ms` - people should use `ThisThread::sleep` instead. `wait_us` and `wait_ns` remain as busy-waits, and appropriate for odd-context use.

Other functions will sleep and are thread-context only. (`ThisThread::sleep`, `Mutex::lock`, `Semaphore::acquire`).

Aim is to reduce confusion about what these calls do, by name adjustment. `wait` implies busy-wait.

Now, we don't quite get there because there are still a couple of "wait"s in the RTOS API - `ConditionVariable::wait` and `EventFlags::wait`, and they obviously sleep, but I don't think there's a better name; cv::wait is totally C/C++ standard and I'd rather not deviate from CMSIS-RTOS either. (`Semaphore::wait` is there now, but I'm intending to make that become `Semaphore::acquire`.

This might be an argument for shifting `wait_us` and `wait_ns` to a new name like `delay_us` and `delay_ns`.  (Would also align us to the Linux kernel, which uses `sleep` and `delay` as its two terms...)

0Grit replied:
Would it hurt to be more explicit? ```delay_execution_ns``` , ```delay_blocking_ns``` ,  ```no_operation_ns```

kjbracey replied:
None of those are immediately grabbing me. "Blocking" often refers to sleeping. "Execution" is saying what? And there aren't necessarily any no-ops involved. Maybe you could include "spin" or "busy"?

bridadan replied:
If I had to make a suggestion for the deprecation message, I'd go with something similar to following. I think it gets to the resolution a bit quicker.

```c++

MBED_DEPRECATED_SINCE("mbed-os-5.13",
                      "'wait' is deprecated in favor of explicit sleep functions. To sleep, 'wait' should be replaced by "
                      "'ThisThread::sleep_for' (C++) or 'thread_sleep_for' (C). If you wish to wait (without sleeping), call "
                      "'wait_us'. 'wait_us' is safe to call from IRQ context.")
```

bridadan commented on platform/internal/mbed_os_timer.cpp:
```diff
@@ -0,0 +1,243 @@
+
+/* (uint32_t)-1 delay is treated as "wait forever" */
+/* This version is tuned for non-RTOS use, where we don't need to return sleep time, and waiting forever is possible */
+void do_timed_sleep_relative_or_forever(uint32_t wake_delay, bool (*wake_predicate)(void *), void *wake_predicate_handle)
```
Warning: this is very much a bike-shedding topic, so I'll endeavor to keep this as short as possible.

To me, the `do_timed_sleep_relative_or_forever` function name is pretty confusing, specifically the `_or_forever` suffix. Is it a relative sleep or a forever sleep? After thinking about it for a bit, I generally assume that I can pass an argument in that makes it a forever sleep (which turns out to be the case), however the name of the function causes enough cognitive dissonance for me to write this comment 😆. Is this a pretty common function name pattern in other libraries?

Other question is how come there isn't an equivalent function for the RTOS implementation? Perhaps having a thread sleep forever doesn't make sense, but I might be missing something else (might be good to add to the commit message?).

My knee-jerk suggested alternative would be have a separate `do_sleep_forever` function OR move the "forever sleep" functionality into either the `do_timed_sleep_relative` or `do_timed_sleep_absolute` function by supplying a constant value (as you have here).

If there's an obvious reason for the current implementation, then please forgive my slowness 😄 But also probably wouldn't hurt to add this reason to the commit message!

kjbracey replied:
This is marked as internal, it's not intended for public use, so I'm not too concerned about naming.

Its API is basically "what we need to implement the RTOS API calls in non-RTOS environment smoothly", rather than being any sort of elegant API concept. I'd maybe redesign it if I felt we needed it to be public API.

I'm trying to be utterly explicit in the name that it accepts the magic wake_delay of `osWaitForever` (`(uint32_t)-1`) . That isn't the case for some of the calls it's alongside. And waiting forever is a clearly distinct case.

bridadan replied:
AH totally missed that this is internal, my bad! In that case, be as verbose as you like :)

bridadan reviewed (commented):
It's obvious a lot of well thought-out work has gone into this, so very nice work @kjbracey-arm. I have a few tweaks and follow up questions below.

In general I don't think I can comment on all the timing changes since I'm not that familiar with them, but I've tried to make helpful suggestions where I can.

kjbracey committed:
rtos/mbed_lib.json: Correct indents

bridadan reviewed (approved):
You've addressed my comments! However I should not be the only person to sign off on this, I didn't really look at any kind of backwards-compatibility questions (I can though if you'd like me to)

bulislaw commented on platform/internal/SysTimer.cpp:
```diff
@@ -0,0 +1,318 @@
+/*
```
We are planning to introduce "include" directories so "internal" shouldn't be necessary. Also it makes it very hard to review if you move a file and modify it in the same commit.

kjbracey replied:
Okay, I can adjust patchset so it moves into platform first, so you'll see a clearer diff on single commit view.

You want it to just go into `platform/SysTimer.[cpp|h]` then? As we don't have separate include directories yet, it seemed wise to otherwise keep it apart from the API headers. (It currently is a bit tucked away in `rtos/TARGET_CORTEX`, so doesn't just visibly sit alongside all the other `rtos/Mutex.h`)

kjbracey replied:
Oh, and note that `mbed_atomic_impl.[c|h]` are already in `/platform/internal` from #10147, so that split has already happened on master.

I don't see any problem with having them split like this. I don't see the logic in keeping private stuff with public now just because there's a reorganisation planned. Is it because you're planning `/platform/SysTimer.h` to be the final private position post-reorg?

0xc0170 replied:
> I don't see any problem with having them split like this. I don't see the logic in keeping private stuff with public now just because there's a reorganisation planned. Is it because you're planning /platform/SysTimer.h to be the final private position post-reorg?

I would leave it as well, but good note about `include` directory. We will move internal one accordingly to the new set of rules. Latest rebase moved it outside, so fixed.

bulislaw commented on rtos/ThisThread.cpp:
```diff
@@ -126,25 +212,43 @@ void ThisThread::sleep_until(uint64_t millisec)
+#if MBED_CONF_RTOS_PRESENT
     osThreadYield();
+#else
+    asm volatile("yield");
```
Is that portable across toolchains?

kjbracey replied:
Ah, nearly. I believe ARMC5 will reject it due to the `volatile`.  Removing the `volatile` is fine for the other toolchains. (It's implied by the lack of any outputs).

bulislaw reviewed (changes requested):
Great work Kevin, I'm ok with most of the changes. For the future reference could you group the related commits in smaller PRs pretty please.

kjbracey committed:
Move SysTimer to platform, for non-RTOS use

kjbracey committed:
Switch SysTimer license to Apache

0xc0170 committed:
VK_RZ_A1H: add mbed_get_a9_tick_irqn

0xc0170 commented on targets/TARGET_RENESAS/TARGET_RZ_A1XX/TARGET_VK_RZ_A1H/device/os_tick_ostm.c:
```diff
@@ -191,5 +191,10 @@ uint32_t OS_Tick_GetOverflow (void)
 }
 
+// Get Cortex-A9 OS Timer interrupt number
+IRQn_ID_t mbed_get_a9_tick_irqn(){
```
I am not certain why we dont use OS_Tick_GetIRQn but have this additional function?

I did not want to clean it up here, so added it for this target as well.

kjbracey replied:
When in non-tickless mode, ARM-A targets need to provide all the `OS_Tick` functions, while mbed core provides it for ARM-M. It's the timer porting layer for RTX.

When in tickless mode, mbed_rtx_idle.c provides all the `OS_Tick` functions, built on the LPTicker SysTimer, for both ARM-A and ARM-M. And for ARM-A (or ARM-M with "No-SysTick"), it needs to know an interrupt number it can trigger in software. That might not be the same interrupt number as the hardware tick interrupt you'd use in tick mode

I didn't think anything had changed in that bit of the structure in this PR...

kjbracey committed:
SysTimer: don't always hold deep sleep lock

Revert back to older behaviour where we hold deep sleep lock only while
timing a sleep. Previous version was a speed optimisation, but broke
some tests.

0xc0170 committed:
SysTimer: irq Cortex A fix C linkage

geky commented on events/equeue/equeue_mbed.cpp:
```diff
@@ -120,27 +123,28 @@ void equeue_mutex_unlock(equeue_mutex_t *m)
 
 // Semaphore operations
-#ifdef MBED_CONF_RTOS_PRESENT
+#ifdef MBED_CONF_RTOS_API_PRESENT
```
Can these two compile-paths be consolidated with the baremetal RTOS APIs?

kjbracey replied:
It's still theoretically possible to make a "bare metal without RTOS API" build, at the moment - that was one of the issues that didn't really get discussed. (Bullet point in description: "bare-metal: requires rtos-api?")

Or we could actually say that "events" requires "rtos-api" in its JSON, and then drop the other case.

The other case here is a bit lighter than rtos-api sleeping in terms of ROM size, but the cost is that you lose the deep sleep latency compensation stuff, which could be problematic. I'm inclined to think that it would be better to drop it and have the only sleep logic in do_sleep.

A possible compile choice that might be worthwhile outside events is lighter-weight versions of the "do_sleep" family. Adding that does increase blinky size - because it now sleeps. I can imagine options for a lighter-weight builds:

* no sleep manager - always just shallow sleep
* Maybe don't even use HAL sleep, just WFI?
* Just spin without timer + sleep/WFI at all - back to the old baremetal `wait_ms` behaviour.

geky commented on events/equeue/equeue_mbed.cpp:
```diff
@@ -177,23 +182,21 @@ static void equeue_sema_timeout(equeue_sema_t *s)
+    while (!*s && ms != 0) {
         sleep();
         core_util_critical_section_exit();
+        __ISB();
```
Relatedly, does this still need to be special? Or can we delegate to the new `do_timed_sleep_absolute` somehow?

kjbracey replied:
You're right, that is a halfway house I hadn't thought of - that is available without the RTOS API.

If we did decide to keep two paths, conditional on RTOS API, then yes, would make sense to use it.

geky reviewed (commented):
Sorry for being late, don't let me block anything. I can create a separate PR to go in after this one if necessary. These are good changes, though it will be interesting to deprecate wait().

Just had some questions over equeue_sema integration with the baremetal APIs.

kjbracey committed:
rtos: NULL -> nullptr

Some build errors seen with NULL not being defined. Rather than
add cstddef includes, switch to C++11 nullptr.

kjbracey committed:
equeue - avoid Kernel::get_ms_count from IRQ

Kernel::get_ms_count is documented as not working from IRQ.

In RTOS builds it can return misleading answers - see
https://github.com/ARM-software/CMSIS_5/issues/625

In non-RTOS builds, it can trigger an assert, as it upsets the
sleep logic.

Modified code is still not ideal - could be improved further if
there was a fast path for "post now" that didn't bother looking
at timers (both at post time and dispatch time).

kjbracey committed:
events: Don't use OsTimer in SysTick builds

Previous fix assumed OsTimer is in use - it is for tickless RTOS builds
and non-RTOS builds, but non-tickless RTOS builds avoid it altogether
and use the SysTick peripheral.

Restore previous behaviour for those builds, and just always read the
tick count - there is no downside to this when ticking.

While in the code, make equeue_tick initialisation explicit. This was
problem if the OsTimer is not yet initialised when th
done while debugging - turns out not to be part of the fix, but it
speeds up the runtime code.

bulislaw reviewed (approved):
Looks good, but please try to submit smaller PRs in the future.
